### ISSUE TITLE
extractkernel for rdc code

### ISSUE THREAD

**Issue by brad-mengchi** [ROCm-Developer-Tools/HIP#2274]
Hi all,

I tried to write a global kernel and a device function and compile device function in a relocatable way(-fgpu-rdc). Then I tried to use extractkernel script to dump the AMD GPU assembly code.

Seems that it fails to dump since there is no .hip_fatbin section after extracting(objdump -h) the executable. So does anyone know which section (names) extractkernel needs to recognize to dump GPU kernel assembly?

P.S.: I am using ROCm 4.0.0 with clang(llvm) 12.

Thanks!

**Comment by ex-rzr** [comment]
As extractkernel is removed from recent versions of ROCm (starting from 4.1) I doubt this will be fixed.
I would recommend you to call hipcc with `-save-temps`.

**Comment by scchan** [comment]
`extractkernel` has been replaced by 2 separate code object tools: `roc-obj-ls` and `roc-obj-extract`
`roc-obj-ls <exe> | roc-obj-extract` provides a similar functionality as `extractkernel`.  The recent version of the HIP programming guide has more details about these 2 tools.
`-save-temps` is another alternative as @ex-rzr has pointed out if you have the source files.

**Comment by brad-mengchi** [comment]
> `extractkernel` has been replaced by 2 separate code object tools: `roc-obj-ls` and `roc-obj-extract`
> `roc-obj-ls <exe> | roc-obj-extract` provides a similar functionality as `extractkernel`. The recent version of the HIP programming guide has more details about these 2 tools.
> `-save-temps` is another alternative as @ex-rzr has pointed out if you have the source files.

Thanks @scchan! But I have some legacy problem that I need to use ROCm 4.0.0 where I think the roc-obj-ls is not available. I think I will try with -save-temps.

**Comment by brad-mengchi** [comment]
> As extractkernel is removed from recent versions of ROCm (starting from 4.1) I doubt this will be fixed.
> I would recommend you to call hipcc with `-save-temps`.

Hi @ex-rzr and @scchan , -save-temps generate a bunch of different files end in .bc, .mcin, .o and .hipfb. Most of them are object file(I think). So I need extra tools to interpret the object files?

**Comment by ex-rzr** [comment]
Among all files there should be files like `*-gfx908.s`, do you have them?

**Comment by brad-mengchi** [comment]
Thanks, @ex-rzr for your patience!

Nop, I do not see files like *-gfx908.s. My case is a little complicated. Here is the repo for my example code: https://github.com/brad-mengchi/llvm-lto. It is mainly derived from HIP-example vectorAdd.

I tried to compile and see whether hipcc can perform LTO(-flto) with manually built LLVM.  I use compile.sh script to compile the files(a.cu and main.cu) where I separately compile and enable LTO(-flto):

>hipcc -fgpu-rdc -c -flto a.cpp -o a.o
>hipcc -fgpu-rdc -c -flto vectoradd_hip.cpp -o vectoradd_hip.o
>hipcc -fgpu-rdc --hip-link -flto -v -save-temps a.o vectoradd_hip.o -o vectoradd_hip.exe

I only dump(-save-temps) at the last step since I would like to see the whole GPU device assembly code to understand whether hipcc(and its underneath LLVM) can perform LTO.

I am not sure whether it is because I tried to perform the separate compilation or tried to perform the LTO(-flto) to an example. Or it is just because no assembly code could be dumped at link time(This might be more possible). That's why I also look for extractkernel, where we can dump the assembly from the binary(but seems extractkernel cannot recognize the binary kernel sections).

Note: to enable LTO for LLVM, I also have to compile the LLVMgold.so plugin libraries.

**Comment by scchan** [comment]
@brad-mengchi   -fgpu-rdc is currently implemented with LTO.  I don't see a .s file with -fgpu-rdc and -save-temps either so that's probably a bug.  You could disassemble the a.out-hip-amdgcn*.lto.o with llvm-objdump --disassemble.

@aaronenyeshi Could you take a look at the -save-temps issue?

**Comment by yxsamliu** [comment]
clang uses lld to do bitcode linking and codegen for -fgpu-rdc. lld -save-temps does not save .s file. However, it saves a.out-*precodegen.bc. You can use /opt/rocm/llvm/bin/llc with that bc file and get .s, which should be close to what you want.

**Comment by brad-mengchi** [comment]
Thanks @yxsamliu, I can dump the llc from the bc file to get the assembly code!

More interesting is that I found seems that llvm can inline the relocatable code(perform the link-time optimization) for the AMDGPU code even without -flto option. @scchan You said that -fgpu-rdc implemented with LTO. Does this mean AMDGPU implemented standalone LTO instead of using LLVM's infrastructure?

I found the relocatable code can be inlined between *.0.2.internalize.bc and *.0.4.opt.s. So I guess AMD GPU pass could perform this inlining after variable internalization and even before (main) optimization?

After all, thanks all so much @yxsamliu @scchan @ex-rzr for your quick and detailed answers!
